### PR TITLE
Hacky Windows support for Electron

### PR DESCRIPTION
**Pull request by cjb**
This PR gets Electron up and running on Windows!  We probably don't want to merge it in this form:

* We don't understand why the net.connect({}) call is needed.
* We can't use socketRoot for splash file storage -- it's a virtual filesystem for named pipes, can't put files there.  So this special cases dataRoot for the splash file only on Windows.  We should consider using dataRoot everywhere, last time around we couldn't because of something to do with ephemeral storage on Linux.
* I do the net.connect({}) in renderer/index.js, but that doesn't cover all renderer windows, it would be better to put it somewhere where *any* new BrowserWindow will run it.  I don't yet know whether it needs to be run in all of them but it might.

### CODE REVIEW

**Comment by cjb** [comment]
@keybase/react-hackers ^

**Comment by cjb** [comment]
@zanderz ^

File: react-native/react/constants/platform.native.desktop.js [hunk 1]
```diff
@@ -19,10 +19,20 @@ const envedPathOSX = {
   prod: 'Keybase'
 }
 
+function buildWin32SocketRoot() {
+  let appdata = process.env.APPDATA
+  // Remove leading drive letter e.g. C:
+  if (/^[a-zA-Z]:/.test(appdata)) {
+    appdata = appdata.slice(2)
+  }
+  return `\\\\.\\pipe\\kbservice\\${appdata}\\Keybase`
```
Comment (zanderz) [inline react-native/react/constants/platform.native.desktop.js]: keybase.exe and kbfs allow for an environment and/or command line variable to replace this. Also, currently it is different in develop vs. production.

**Comment by zanderz** [comment]
LGTM, whatever works, until electron gets worked out. I will try to look into what happens behind net.connect({}) on Windows

File: desktop/app/splash.js [hunk 2]
```diff
@@ -1,9 +1,9 @@
 // Checks to see if we've launched before (we touch started.txt). If not, write the file and open a browser
-import {socketRoot} from '../../react-native/react/constants/platform.native.desktop'
+import {socketRoot, dataRoot} from '../../react-native/react/constants/platform.native.desktop'
 import {helpURL} from '../../react-native/react/constants/urls'
 import path from 'path'
 
-const filePath = path.join(socketRoot, 'started.txt')
+const filePath = path.join(process.platform === 'win32' ? dataRoot : splashRoot, 'started.txt')
```
Comment (chrisnojima) [inline desktop/app/splash.js]: we should just expose a different variable from the constants that does the check there.

File: desktop/renderer/index.js [hunk 3]
```diff
@@ -41,6 +42,16 @@ class Keybase extends Component {
   constructor () {
     super()
 
+    // This net.connect() is a heinous hack.
+    //
+    // On Windows, but *only* in the renderer thread, our RPC connection
+    // hangs until other random net module operations, at which point it
+    // unblocks.  Could be Electron, could be a node-framed-msgpack-rpc
+    // bug, who knows.
+    if (process.platform === 'win32') {
```
Comment (chrisnojima) [inline desktop/renderer/index.js]: if we do need to keep this hack maybe let's just isolate it in a different module and not in here. So we can do like
```
import windowsHack from '../../react-native/react/native/windows-hack'
```
then just windowsHack()

File: react-native/react/constants/platform.native.desktop.js (hunk 1 above)
Comment (chrisnojima) [inline react-native/react/constants/platform.native.desktop.js]: also we can use electron to get a bunch of this path stuff (maybe we should cleanup the constants file)

https://github.com/atom/electron/blob/master/docs/api/app.md#appgetpathname

**Pull request merged**